### PR TITLE
[Refac] PyQHamiltonianEvolution - Added the case for diagonal Hamiltonian

### PR DESCRIPTION
**Pull request by neogyk**
Refactored the `PyQHamiltonianEvolution`. Added the case for diagonal Hamiltonian  #368

### CODE REVIEW

**Comment by dominikandreasseitz** [comment]
hi @neogyk , thanks for participating! can you add a test case under `tests/backends/pyq/test_quantum_pyq.py`

File: qadence/backends/pyqtorch/convert_ops.py [hunk 1]
```diff
@@ -313,6 +316,19 @@ def __init__(
...
+            and not block.is_parametric
+            and block.generator._is_diag_pauli
+        ):
+            hmat = block_to_diagonal(
+                block.generator, qubit_support=self.qubit_support, use_full_support=False
+            )
+
+            hmat = hmat.permute(1, 2, 0)
+            self.register_buffer("hmat", hmat)
+            self._hamiltonian = lambda self, values: self.hmat
```
Comment (jpmoutinho) [inline qadence/backends/pyqtorch/convert_ops.py:330]: Note that here `self.hmat` should now be a single vector instead of a matrix, so I suspect you still need to change the code when this function is called to account for that. You should look at the `unitary` and `_unitary` methods, and the `_evolve_diag_operator` function.

As for the test, like @dominikandreasseitz, first let's see if all current tests pass. It might be enough for this change, this it is just a refactoring of something that was hopefully already tested. But I will take a look to double-check.

File: qadence/backends/pyqtorch/convert_ops.py [hunk 2]
```diff
@@ -313,6 +316,19 @@ def __init__(
             self.register_buffer("hmat", hmat)
             self._hamiltonian = lambda self, values: self.hmat
 
+        if (
```
Comment (jpmoutinho) [inline qadence/backends/pyqtorch/convert_ops.py:319]:
```suggestion
        elif (
```

**Review by jpmoutinho — COMMENTED** [review]
Thanks for the contribution @neogyk!

File: qadence/backends/pyqtorch/convert_ops.py (hunk 1 above)
Comment (neogyk) [reply qadence/backends/pyqtorch/convert_ops.py:330]: @jpmoutinho Thank you, I will make the corrections

**Comment by jpmoutinho** [comment]
Hi @neogyk, let us know if you have any questions on how to complete this :)

**Comment by neogyk** [comment]
Hello, @jpmoutinho. I would like to clarify, what is the structure of the input of  `unitary` method of `PyQHamiltonianEvolution`. I found that it's require to store the value of the time variable in the , and accessible by the parameter name, which is in the default case is [uuids](https://github.com/pasqal-io/qadence/blob/5865cf3cbfa03e3e341034ab02cac2bd04bc14c7/qadence/backend.py#L87). 
Also, the 'PyQHamiltonianEvolution' assumes that the input block hamiltonian is batches or time evolution is batched.
Could you please what is the input can be used for testing the `PyQHamiltonianEvolution `.
Currently I am trying to test it with the following code:
  `qubit_support = (0,)

    n_qubits = 2

    block = HamEvo(generator=zz_hamiltonian(n_qubits), parameter=torch.tensor(100))

    config = PyqConfig()

    pyq_ham_evo = convert_ops.PyQHamiltonianEvolution(qubit_support, n_qubits, block, config)

    val_dict = {config.get_param_name(block)[0]:torch.tensor([100])}

    pyq_ham_evo.unitary(val_dict)

    state = random_state(n_qubits=2)

    result = pyq_ham_evo.forward(state=state, values=val_dict)

**Comment by jpmoutinho** [comment]
> Hello, @jpmoutinho. I would like to clarify, what is the structure of the input of `unitary` method of `PyQHamiltonianEvolution`. I found that it's require to store the value of the time variable in the , and accessible by the parameter name, which is in the default case is [uuids](https://github.com/pasqal-io/qadence/blob/5865cf3cbfa03e3e341034ab02cac2bd04bc14c7/qadence/backend.py#L87). Also, the 'PyQHamiltonianEvolution' assumes that the input block hamiltonian is batches or time evolution is batched. Could you please inform me what is the input can be used for testing the `PyQHamiltonianEvolution `. Currently I am trying to test it with the following code:
> 
> ```
> qubit_support = (0,)
> 
> n_qubits = 2
> 
> block = HamEvo(generator=zz_hamiltonian(n_qubits), parameter=torch.tensor(100))
> 
> config = PyqConfig()
> 
> pyq_ham_evo = convert_ops.PyQHamiltonianEvolution(qubit_support, n_qubits, block, config)
> 
> val_dict = {config.get_param_name(block)[0]:torch.tensor([100])}
> 
> pyq_ham_evo.unitary(val_dict)
> 
> state = random_state(n_qubits=2)
> 
> result = pyq_ham_evo.forward(state=state, values=val_dict)
> ```

Hi @neogyk, I suggest you test it with the following code:

```
from qadence import X, Z, HamEvo, random_state, run

# Diagonal Hamiltonian:
gen_diag = Z(0) + Z(1) 
# Adding and cancelling an X term will make it count as non-diagonal:
gen_non_diag = Z(0) + Z(1) + X(0) - X(0) 

state = random_state(n_qubits=2)

state_from_diag = run(HamEvo(gen_diag, torch.tensor([1.0])))
state_from_nondiag = run(HamEvo(gen_non_diag, torch.tensor([1.0])))
```

Hopefully that will serve as a good basis for you to figure out if you are doing it correctly.

I'm afraid I don't have a lot of time today to look at this MR, I will try to take a look later today.

**Comment by dominikandreasseitz** [comment]
closing as discussed with @jpmoutinho

**Comment by neogyk** [comment]
Dear @dominikandreasseitz , is this issue is still actual?
I made the changes in the work of  [`PyQHamiltonianEvolution`](https://github.com/neogyk/qadence/commit/71847287ac714363c7f716a420976a80f6ebdcfe)

**Comment by jpmoutinho** [comment]
Hi @neogyk. Unfortunately since the UnitaryHack concluded last week we assumed you wouldn't be pursuing this further, given that you are no longer eligible for the reward.

At this moment we are also concluding some other refactoring that is related to this part of the codebase, so I am afraid it wouldn't be a good opportunity to re-open this MR :(

Either way, thanks a lot for your interest and effort, we really appreciate the initiative! If you want to stay involved in the Pasqal open-source ecosystem, I would suggest that you take a look at PyQTorch (https://github.com/pasqal-io/pyqtorch). It's an overall simpler library and we are currently in the process of moving certain things there.

**Comment by neogyk** [comment]
@jpmoutinho, thank you. I will take a look at the PyQTorch.

**Comment by neogyk** [comment]
@dominikandreasseitz , @jpmoutinho, I received the information, that the deadline for the review was extended till 26th of June.

**Comment by jpmoutinho** [comment]
Hi @neogyk, I see, we got that information now as well. I was going to re-open this MR but it seems that you have force-pushed or recreated the branch, and thus github does not allow me.

Feel free to re-open a new MR.

**Comment by neogyk** [comment]
Hello, @jpmoutinho. I created the new MR #478
